### DISCUSSION TITLE
Generate mesh for applying periodic boundary conditions in Neper

### DISCUSSION THREAD

**Discussion by ashokambig** [neperfepx/neper#581] [Q&A] [answered]
Hello All,

I want to create a cube microstructure and mesh it with tetrahedral elements to run a simulation in Abaqus.
In the simulation I have to apply periodic boundary conditions. To apply periodic boundary conditions it is better if I have similar mesh on the opposite faces of the cube --> Number of elements and nodes on the opposite sides of the cube must be same.
I have the below code for generating mesh.
Is there a way to create such a mesh in Neper by putting a mesh control feature in the below code?

Regards
Ashok

code:
*****************************************
neper -T -n 5000 -domain "cube(20,20,20)" -reg 1 -o abq_input
neper -V abq_input.tess -datacellcol id -print imag_1
neper -M abq_input.tess -elttype tet -faset none -format inp

**Comment by rquey** [comment, marked as answer]
No. This is the type of periodic polycrystal that Neper can generate and mesh:

```
neper -T -n 1000 -morpho gg -per 1
neper -M n1000-id1.tess -pl 100
neper -V n1000-id1.tess,n1000-id1.msh -showelt1d all -dataelt1drad 0.0025 -dataelt3dedgerad 0.001 -cameracoo x:5:z -cameraprojection orthographic  -print img
```

![image](https://user-images.githubusercontent.com/2659376/195840279-385b9471-b710-4455-99a3-8e966c0baebd.png)

**Reply by Mahdieh-karami** [reply]
Hi, Is it possible to generate smooth edges ?

**Reply by rquey** [reply]
What do you mean by this?

**Reply by Mahdieh-karami** [reply]
![image](https://github.com/user-attachments/assets/e2bc88b6-1a41-4679-b96b-648b2f67e067)

I mean something like the picture. Grains on the border will be completed If we put the edges together(the grains have been cut that I showed in the same color belong to one grain), so it creates a periodic shape, unlike what Neper makes the edges or in 3D the plane on the boundary are smooth.

**Reply by rquey** [reply]
See this code: https://www.softxjournal.com/article/S2352-7110(24)00282-6/fulltext.
I've not tested it nor can provide any type of assistance though.

**Reply by Mahdieh-karami** [reply]
Thank you.

**Reply by Mahdieh-karami** [reply]
Hi, 
I used Neper and made polycrystal like image using DCT scan data,  now I should use mosaic to have periodic  rectilinear but I could not make geo file . could you please help me . this is code that I have to use :python3 -m neper_mosaic n78-id1.geo myRVE_periodic.msh -v --show_gui
![image](https://github.com/user-attachments/assets/95bcbf59-6392-4d74-8e81-68d4901ee7cf)

**Reply by dilekguzel** [reply]
Hello,
you can get geo format from Neper by adding -format geo, such that you will get not .tess but .geo files
For example:
 -T -n 20 -id 1 -morpho "gg,diameq:lognormal(0.6,0.05)" -periodicity all -format geo

Best regards
Dilek
